server/console_account.go

```diff
@@ -208,6 +256,188 @@ func (s *ConsoleServer) GetWalletLedger(ctx context.Context, in *console.Account
+			}
+			userID = &uid
+		}
+
+		if userID != nil {
+			// Looking up a single specific tombstone.
+			var createTime pgtype.Timestamptz
+			err := s.db.QueryRowContext(ctx, "SELECT create_time FROM user_tombstone WHERE user_id = $1", *userID).Scan(&createTime)
+			if err != nil {
+				if err == sql.ErrNoRows {
+					return &console.AccountList{
+						TotalCount: countAccounts(ctx, s.logger, s.db),
+					}, nil
+				}
+				s.logger.Error("Error looking up user tombstone.", zap.Any("in", in), zap.Error(err))
+				return nil, status.Error(codes.Internal, "An error occurred while trying to list users.")
+			}
+
+			return &console.AccountList{
+				Users: []*api.User{
+					{
+						Id:         in.Filter,
+						UpdateTime: &timestamp.Timestamp{Seconds: createTime.Time.Unix()},
+					},
+				},
+				TotalCount: countAccounts(ctx, s.logger, s.db),
+			}, nil
+		}
+
+		query := "SELECT user_id, create_time FROM user_tombstone LIMIT 50"
+
+		rows, err := s.db.QueryContext(ctx, query)
+		if err != nil {
+			s.logger.Error("Error querying user tombstones.", zap.Any("in", in), zap.Error(err))
+			return nil, status.Error(codes.Internal, "An error occurred while trying to list users.")
+		}
+
+		users := make([]*api.User, 0, 50)
+
+		for rows.Next() {
+			var id string
+			var createTime pgtype.Timestamptz
+			if err = rows.Scan(&id, &createTime); err != nil {
+				_ = rows.Close()
+				s.logger.Error("Error scanning user tombstones.", zap.Any("in", in), zap.Error(err))
+				return nil, status.Error(codes.Internal, "An error occurred while trying to list users.")
+			}
+
+			users = append(users, &api.User{
+				Id:         id,
+				UpdateTime: &timestamp.Timestamp{Seconds: createTime.Time.Unix()},
+			})
+		}
+		_ = rows.Close()
+
+		return &console.AccountList{
+			Users:      users,
+			TotalCount: countAccounts(ctx, s.logger, s.db),
+		}, nil
+	}
+
+	if in.Filter != "" {
+		_, err := uuid.FromString(in.Filter)
+		// If the filter is not a valid user ID treat it as a username instead.
+
+		var query string
+		params := []interface{}{in.Filter}
+		if err != nil {
+			query = "SELECT id, username, display_name, avatar_url, lang_tag, location, timezone, metadata, apple_id, facebook_id, facebook_instant_game_id, google_id, gamecenter_id, steam_id, edge_count, create_time, update_time FROM users WHERE username = $1"
+		} else {
+			query = "SELECT id, username, display_name, avatar_url, lang_tag, location, timezone, metadata, apple_id, facebook_id, facebook_instant_game_id, google_id, gamecenter_id, steam_id, edge_count, create_time, update_time FROM users WHERE id = $1"
+		}
+
+		if in.Banned {
+			query += " AND disable_time <> '1970-01-01 00:00:00 UTC'"
+		}
+
+		rows, err := s.db.QueryContext(ctx, query, params...)
+		if err != nil {
+			s.logger.Error("Error querying users.", zap.Any("in", in), zap.Error(err))
+			return nil, status.Error(codes.Internal, "An error occurred while trying to list users.")
+		}
+
+		users := make([]*api.User, 0, 2)
```

Comment:
why allocate a capacity for 2 users, shouldn't the filter only return one?

Reply:
The code requires revising when it is implemented properly.